synapse/storage/data_stores/main/events_worker.py

```diff
@@ -580,8 +583,49 @@ def _get_events_from_db(self, event_ids, allow_rejected=False):
                 # of a event format version, so it must be a V1 event.
                 format_version = EventFormatVersions.V1
 
-            original_ev = event_type_from_format_version(format_version)(
+            room_version_id = row["room_version_id"]
+
+            if not room_version_id:
+                # this should only happen for out-of-band membership events
+                if not internal_metadata.get("out_of_band_membership"):
+                    logger.warning(
+                        "Room %s for event %s is unknown", d["room_id"], event_id
+                    )
+                    continue
+
+                # take a wild stab at the room version based on the event format
+                if format_version == EventFormatVersions.V1:
+                    room_version = RoomVersions.V1
+                elif format_version == EventFormatVersions.V2:
+                    room_version = RoomVersions.V3
+                else:
+                    room_version = RoomVersions.V5
```

Comment:
Is the idea that for out of band memberships after this commit will have a room version stored? I.e. we should never get here once for later room versions?

Reply:
well, out-of-band-memberships since #6983. But yes.